File: src/main/java/leetcode/ValidParentheses.java
package leetcode;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.Map;

public class ValidParentheses {
  public boolean isValid(String s) {
    Map<Character, Character> map = Map.of(')', '(', '}', '{', ']', '[');
    Deque<Character> stack = new ArrayDeque<>();
    for (char c : s.toCharArray()) {
      if (!map.containsKey(c)) {
        stack.push(c);
      } else if (stack.isEmpty() || stack.pop() != map.get(c)) {
        return false;
      }
    }
    return stack.isEmpty();
  }
}

File: leetcode-kt/src/main/kotlin/com/giwankim/leetcode/ValidParentheses.kt
package com.giwankim.leetcode

import java.util.ArrayDeque
import java.util.Deque

class ValidParentheses {
    fun isValid(s: String): Boolean {
        val map: Map<Char, Char> = mapOf(')' to '(', '}' to '{', ']' to '[')
        val stack: Deque<Char> = ArrayDeque()
        for (c in s) {
            if (!map.containsKey(c)) {
                stack.push(c)
            } else if (stack.isEmpty() || stack.pop() != map[c]) {
                return false
            }
        }
        return stack.isEmpty()
    }
}
